### ISSUE TITLE
[medgan-pr-prep] T3: Refactor MedGAN to proper BaseModel, add train_model() and synthesize_dataset()

### ISSUE THREAD

**Issue by jalengg** [jalengg/PyHealth#39]
## Goal

Remove the DummyWrapper hack and make `MedGAN` a proper PyHealth 2.0 BaseModel with `train_model()` and `synthesize_dataset()`.

### Remove

- `DummyWrapper` inner class in `__init__`
- `from_binary_matrix()` classmethod (old API, not PyHealth 2.0)
- `from_count_matrix()` classmethod (old API)
- `_build_global_vocab()` (processor handles this now)
- `_extract_features_from_batch()` internal helper (replaced by processor)
- `train_step()` raw method (replaced by `train_model()`)

### Modify `__init__`

```python
def __init__(self, dataset, latent_dim=128, hidden_dim=128,
             autoencoder_hidden_dim=128, discriminator_hidden_dim=256,
             minibatch_averaging=True, save_dir="./save/", **kwargs):
    super().__init__(dataset)          # real dataset, no DummyWrapper
    visits_processor = dataset.input_processors["visits"]
    self.input_dim = visits_processor.size()   # use .size(), not .vocab_size()
    # build components using self.input_dim
    ...
```

### Add `train_model()`

Two-phase training:
1. **Phase 1**: Pretrain autoencoder — BCE loss (`nn.BCELoss`), Adam optimizer
2. **Phase 2**: GAN training — standard BCE (NOT WGAN), alternate D/G updates

```python
def train_model(self, train_dataset, val_dataset=None, 
                ae_epochs=100, gan_epochs=100, batch_size=32,
                ae_lr=0.001, gan_lr=0.0002):
    # Phase 1: Autoencoder pretraining
    # Phase 2: GAN training
    # Save checkpoint via save_model()
```

### Add `synthesize_dataset()`

```python
def synthesize_dataset(self, num_samples, random_sampling=True) -> List[Dict]:
    # 1. Sample noise: torch.randn(num_samples, latent_dim)
    # 2. Generator: noise → hidden
    # 3. Autoencoder decoder: hidden → binary vector (sigmoid)
    # 4. Threshold at 0.5 → binary
    # 5. Decode binary vector back to code strings using processor vocab
    # return [{"patient_id": "synthetic_N", "visits": [code, ...]}]
```

### Make `forward()` raise `NotImplementedError`

```python
def forward(self, **kwargs):
    raise NotImplementedError("Use train_model() for training and synthesize_dataset() for generation.")
```

### Add `save_model()` / `load_model()`

```python
def save_model(self, path):
    torch.save({
        "autoencoder": self.autoencoder.state_dict(),
        "generator": self.generator.state_dict(),
        "discriminator": self.discriminator.state_dict(),
        "input_dim": self.input_dim,
        "latent_dim": self.latent_dim,
    }, path)

@classmethod
def load_model(cls, path, dataset, **kwargs):
    checkpoint = torch.load(path, map_location="cpu")
    model = cls(dataset, **kwargs)
    model.autoencoder.load_state_dict(checkpoint["autoencoder"])
    model.generator.load_state_dict(checkpoint["generator"])
    model.discriminator.load_state_dict(checkpoint["discriminator"])
    return model
```

## GAN loss details (IMPORTANT: NOT WGAN)

- Discriminator loss: `BCE(D(real), 1) + BCE(D(fake.detach()), 0)`
- Generator loss: `BCE(D(fake), 1)`
- No weight clipping, no gradient penalty

## Acceptance criteria

- `MedGAN(dataset)` works with a `SampleDataset` that has `input_processors["visits"]`
- `forward()` raises `NotImplementedError`
- `train_model()` runs without error (1 epoch smoke test)
- `synthesize_dataset(5)` returns 5 dicts, each with `"patient_id"` and `"visits"` (flat list of str)
- `save_model()` / `load_model()` round-trip preserves weights

**Comment by jalengg** [comment]
## Implementation notes from source analysis

### Current state of medgan.py (before refactor)

The source file has 447 lines with these methods that need attention:

**Keep (but refactor):**
- `MedGANAutoencoder` — solid, keep as-is (linear encoder+decoder, binary/count modes)
- `MedGANGenerator` — solid, keep as-is (2-layer MLP with residual connections, identical to CorGAN)
- `MedGANDiscriminator` — solid, keep as-is (3-layer with sigmoid + minibatch averaging)
- `pretrain_autoencoder()` — good logic, fold into `train_model()` Phase 1 or keep as internal `_pretrain_autoencoder()`
- `train_step()` — good logic, fold into `train_model()` Phase 2 loop
- `generate()` — good logic, used internally by `synthesize_dataset()`
- `sample_transform()` — keep as internal helper for thresholding
- `_init_weights()` — keep as-is

**Remove entirely:**
- `DummyWrapper` inner class in `__init__` — replaced by real `super().__init__(dataset)`
- `from_binary_matrix()` classmethod — old API, not PyHealth 2.0
- `from_count_matrix()` classmethod — delegates to `from_binary_matrix()`, both gone
- `_build_global_vocab()` — processor handles vocabulary now
- `_extract_features_from_batch()` — batch from DataLoader will already have `"visits"` tensor from processor

**Modify:**
- `__init__` — remove DummyWrapper, use `dataset.input_processors["visits"].size()` for `input_dim`
- `forward()` — raise `NotImplementedError` (current implementation does actual computation)

### Critical dimension constraint

`MedGANGenerator` outputs a tensor of shape `(batch, autoencoder_hidden_dim)` which feeds directly into `MedGANAutoencoder.decode()` expecting input of `(batch, hidden_dim)`. Therefore: **`generator hidden_dim` MUST equal `autoencoder_hidden_dim`**. The current code enforces this by passing `autoencoder_hidden_dim` as the hidden_dim for the generator. This constraint must be documented and preserved.

### Two-phase training order

Phase 1 (AE pretrain): `optimizer = Adam(autoencoder.parameters())`, BCE loss, `ae_epochs` epochs.
Phase 2 (GAN): separate optimizers for G and D, standard BCE, `gan_epochs` epochs. During Phase 2, the autoencoder's decoder is used as the final output layer of the generator pipeline (noise → generator → AE.decode() → binary vector).

### `synthesize_dataset()` decode step

After `generate()` returns a `(n_samples, input_dim)` float tensor, need to:
1. Threshold at 0.5 → binary
2. For each row, find indices where value == 1
3. Look up code strings from processor vocab: `processor.idx2token[idx]` or equivalent
4. Return `[{"patient_id": f"synthetic_{i}", "visits": [code, ...]}]`

Need to verify the processor's reverse-lookup API. Check `MultiHotProcessor` in `pyhealth/processors/` for the correct attribute name (`idx2token`, `idx_to_token`, or similar).

### `save_model()` / `load_model()` checkpoint structure

```python
{
    "autoencoder": self.autoencoder.state_dict(),
    "generator": self.generator.state_dict(),
    "discriminator": self.discriminator.state_dict(),
    "input_dim": self.input_dim,
    "latent_dim": self.latent_dim,
    "hidden_dim": self.hidden_dim,
    "autoencoder_hidden_dim": autoencoder_hidden_dim,  # needed for load_model to reconstruct
}
```

### BCE loss — NOT WGAN

Discriminator: `F.binary_cross_entropy(D(real), ones) + F.binary_cross_entropy(D(fake.detach()), zeros)`
Generator: `F.binary_cross_entropy(D(fake), ones)`
NO weight clipping. NO gradient penalty. This is standard GAN.

**Comment by jalengg** [comment]
T3 complete: Refactored MedGAN to proper BaseModel. Removed DummyWrapper, from_binary_matrix, from_count_matrix. Added train_model() (two-phase: AE pretrain + BCE GAN), synthesize_dataset(), save_model/load_model. forward() raises NotImplementedError. Commit: 202fac8e
